packages/babel-plugin-transform-classes/test/fixtures/spec/name-collisions-with-class-ref-in-method/output.js

```diff
@@ -44,19 +44,19 @@ var Baz = /*#__PURE__*/function () {
   babelHelpers.createClass(Baz, [{
     key: "method",
     value: function method() {
-      var _Baz = /*#__PURE__*/function () {
-        function _Baz() {
-          babelHelpers.classCallCheck(this, _Baz);
+      var _Baz2 = /*#__PURE__*/function () {
+        function _Baz2() {
+          babelHelpers.classCallCheck(this, _Baz2);
         }
 
-        babelHelpers.createClass(_Baz, [{
+        babelHelpers.createClass(_Baz2, [{
           key: "f",
           value: function f() {
             var Baz = 1;
             return Baz;
           }
         }]);
-        return _Baz;
+        return _Baz2;
       }();
 
       return babelHelpers.get(babelHelpers.getPrototypeOf(Baz.prototype), "method", this).call(this, _Baz);
```

Comment:
`_Baz` is undefined now. Maybe `environmentVisitor ` is not applicable here?

```js
// environmentVisitor should be used when traversing the whole class and not for specific class elements/methods.
// For perf reasons, the environmentVisitor will be traversed with `{ noScope: true }`, which means `path.scope` is undefined.
// Avoid using `path.scope` here
export const environmentVisitor = {}
```

Reply:
It was a bug anyway, since it should have been

```js
babelHelpers.get(babelHelpers.getPrototypeOf(Baz.prototype), "method", this).call(this, Baz);
```

I think.

Reply:
I think it's a bug in `path.scope.rename()`: even if it renames `Bar` to `_Bar`, it doesn't properly remove `Bar` from the current scope so it will rename it again.

I added some logging around `path.scope.rename()`:
```js
        console.log("================== RENAME", refName, "===============");
        console.log("Name:", path.scope.getOwnBinding(refName).identifier.name);
        console.log("Type:", path.scope.getOwnBinding(refName).path.type);
        console.log("Bindings:", Object.keys(path.scope.bindings));
        console.log(path.toString());

        path.scope.rename(refName);

        console.log("=================================");
        console.log("Bindings:", Object.keys(path.scope.bindings));
        console.log(path.toString());
        console.log("================ DONE =================");
```

And when transforming this code:
```js
class Baz {
  method() {
    class Baz {
      f() {
        let Baz = 1;
        return Baz;
      }
    }

    return super.method(Baz)
  }
}
```

it logs this:
```js
================== RENAME Baz ===============
Name: Baz
Type: ClassDeclaration
Bindings: [ 'Baz' ]
{
  class Baz {
    f() {
      let Baz = 1;
      return Baz;
    }

  }

  return super.method(Baz);
}
=================================
Bindings: [ '_Baz' ]
{
  class _Baz {
    f() {
      let Baz = 1;
      return Baz;
    }

  }

  return super.method(_Baz);
}
================ DONE =================
================== RENAME Baz ===============
Name: _Baz
Type: ClassDeclaration
Bindings: [ 'Baz' ]
class _Baz {
  f() {
    let Baz = 1;
    return Baz;
  }

}
=================================
Bindings: [ 'Baz' ]
class _Baz2 {
  f() {
    let Baz = 1;
    return Baz;
  }

}
================ DONE =================
```

As you can see, the second time it's "renaming Baz" even if it's name is actually already `_Baz`, causing problems.

This is a bug in `@babel/traverse` that will eventually need to be fixed, but we can work it around like this:
```js
      const binding = path.scope.getOwnBinding(refName);
      if (binding && binding.identifier.name === refName) {
        path.scope.rename(refName);
      }
```

With a comment in the code explaining why we need the workaround (you can just link to this discussion).

Reply:
The `@babel/traverse` bug is probably because class declarations introduce two bindings: one internal one and one external one.
When renaming a class `scope.rename()` should first convert `class A {}` to `let A = class A {}`, and then only rename the internal or external binding.

Reply:
🤔 If this is class-specific, we can just skip class-like nodes.

Reply:
And it's working now!